src/nodelets/obstacles_detection.cpp

```diff
@@ -168,14 +159,144 @@ class ObstaclesDetection : public nodelet::Nodelet
 			if(obstaclesPub_.getNumSubscribers())
 			{
 				sensor_msgs::PointCloud2 rosCloud;
-				pcl::toROSMsg(*obstaclesCloud, rosCloud);
+				pcl::toROSMsg(*originalCloud, rosCloud);
 				rosCloud.header.stamp = cloudMsg->header.stamp;
 				rosCloud.header.frame_id = frameId_;
 
 				//publish the message
 				obstaclesPub_.publish(rosCloud);
 			}
+			return;
+		}
+
+		//Common variables for all strategies
+		pcl::PointCloud<pcl::PointXYZ>::Ptr hypotheticalGroundCloud(new pcl::PointCloud<pcl::PointXYZ>);
+		pcl::PointCloud<pcl::PointXYZ>::Ptr obstaclesCloud(new pcl::PointCloud<pcl::PointXYZ>);
+		pcl::IndicesPtr ground, obstacles;
+		pcl::PointCloud<pcl::PointXYZ>::Ptr groundCloud(new pcl::PointCloud<pcl::PointXYZ>);
+
+		ros::Time lasttime = ros::Time::now();
+
+		if (!simpleSegmentation_ && !optimizeForCloseObject_){
+			//This is the default strategy
+			//The cloud is divided into two pointcloud based on reported Z and the position of the camera.
+			// One is the hypothetical ground cloud and the other one is the obstacles pointcloud.
+			//The algorithm then extracts (and remove) from the hypothetical ground cloud the detected obstacles,
+			//and add them to the obstacles pointcloud
+			originalCloud = rtabmap::util3d::transformPointCloud(originalCloud, localTransform);
+			hypotheticalGroundCloud = rtabmap::util3d::passThrough(originalCloud, "z", std::numeric_limits<int>::min(), maxFloorHeight_);
```

Comment:
These first two steps should be common to the three strategies, move them before the conditional